### DISCUSSION TITLE
Discontinuation of Map Store: Implications

### DISCUSSION THREAD

**Discussion by dominikschlosser** [keycloak/keycloak#24052] [Archive]
Hi Keycloak-Team,

the discontinuation of map store saddens us, as we thought it was a great way to increase flexibility of Keycloak's storage layer. It also raises a few questions:

- Where do you make the cut when removing the map store? 
- Specifically: Will the DatastoreProvider-SPI remain as a means to implement our own storage (we use it to replace certain storage and cache areas with a Apache Cassandra implementation: https://github.com/opdt/keycloak-cassandra-extension)

Best regards
Dominik

**Comment by mhajas** [comment]
> * Where do you make the cut when removing the map store?

We have some plans and we will start creating Github issues soon on this. 

> * Specifically: Will the DatastoreProvider-SPI remain as a means to implement our own storage (we use it to replace certain storage and cache areas with a Apache Cassandra implementation: https://github.com/opdt/keycloak-cassandra-extension)

Changes to the `KeycloakSession` we performed will most likely stay as is. We haven't discussed DatastoreProvider yet though. Can you please elaborate more on why `DatastoreProvider` is necessary for you? Is it because you want to avoid both the caching and the managers layer? If yes, do you have your own caching? Do you use LDAP or some other federation?

**Comment by sguilhen** [comment]
> We have some plans and we will start creating Github issues soon on this.

Here's a link to the Epic that will track the removal of the Map Store: https://github.com/keycloak/keycloak/issues/24111

**Comment by dominikschlosser** [comment]
>  Is it because you want to avoid both the caching and the managers layer? 

Excactly! We use Apache Cassandra for persistent storage (users, roles, etc.) as well as temporary data like sessions. Apache Cassandra is fast enough to make any additional caching obsolete. It also solves Cross-DC for us. We are very happy with running Keycloak this way, without all the hassle Infinispan brings with it. 

Keeping DatastoreProvider-SPI would allow us and others to replace the whole storage layer including caching. The added complexity of this indirection should be worth the added flexibility in our opinion :-)

> Do you use LDAP or some other federation?

We do not.

**Reply by mhajas** [reply]
Thanks @dominikschlosser. We will consider your use case when/if discussing `DatastoreProvider` removal.

**Comment by ynojima** [comment]
Allow me to raise another question regarding discontinuation of Map Store.
I heared it is a key element to realize [Zero downtime upgrade](https://github.com/keycloak/keycloak/issues/14455).
Is that affected? Do you have another plan to realize Zero downtime upgrade?

**Reply by mhajas** [reply]
@ynojima You are correct. Map storage was the key element for Zero downtime upgrades and delivery of this feature is affected by this decision. 

Full zero downtime upgrades are not on our roadmap at the moment but we discussed the possibility of having some of the releases (for example some of minor releases) which do not require full Keycloak cluster restarts. However, this still requires some investigation on what are the factors that play a role in deciding whether a release brings some changes that break no-downtime upgradability. This may turn out to be more difficult than we think.

**Reply by ynojima** [reply]
Thank you for clarifying this!

**Comment by jonkoops** [comment]
Please reply in existing threads rather than starting new ones.

**Comment by opdt** [comment]
Hey @mhajas and others,
we migrated our Apache Cassandra Datastore implementation to Keycloak 23 and thus "legacy storage". 
It works but the result is somewhat messy:

- In order to disable infinispan (which we replace with Cassandra) we need not only to provide our own `DatastoreProvider` (which is expected) but override the default provider for `userSessions`, `authenticationSessions`, `loginFailures` and `singleUseObjects` as well (since otherwise the Infinispan-default-implementations are loaded and try to initialize; this was disabled by map-storage feature-flag before)
- There currently are no implementations of `globalLock`, `stickySessionEncoder`, `publicKeyStorage` outside of the Infinispan-implementations. These SPIs are also not part of `Datastore` which makes the handling kind of awkward.
- In order to disable Infinispan, one has to disable two `connectionsInfinispan`-implementations: `default` and `quarkus`
- We basically need to override providers by using the same ID and a higher order to prevent the original implementations (LegacyDatastore, Infinispan-implementations, ...) from registering themselves as event-listeners. This seems like a fundamental problem when overriding provider-implementations since you not only have to make sure to that your provider is used but you also have to prevent that `init()` is called for other provider-implementations you are trying to override...

The biggest problem seems to be that `LegacyDatastoreProvider` is referenced at multiple points in the code in order to use the `*localStorage`- / `*storageManager`-methods which are not part of `DatastoreProvider` and only exist in `LegacyDatastoreProvider`.
Thus we opted to extend `LegacyDatastoreProvider` instead of implementing `DatastoreProvider`.
See also https://github.com/keycloak/keycloak/pull/25572 for an attempt to fix this.

We hope you find this feedback valuable and continue to support our way to replace the default Datastore-implementation like this. If we can help to improve some of these points, please let us know.

**Reply by mhajas** [reply]
Thank you for the thorough explanation @opdt!

> * In order to disable infinispan (which we replace with Cassandra) we need not only to provide our own `DatastoreProvider` (which is expected) but override the default provider for `userSessions`, `authenticationSessions`, `loginFailures` and `singleUseObjects` as well (since otherwise the Infinispan-default-implementations are loaded and try to initialize; this was disabled by map-storage feature-flag before)

If I understand correctly, you are also disabling the caching layer, is it correct? The current store is quite coupled with the caching layer and it is not properly tested to run it without it. On the other hand, I support the idea of having it more flexible so we would be happy to see how this setup works for you, especially about the performance. Rewriting the providers you mentioned is must have to avoid Infinispan as we currently don't have any other implementations. 

> * There currently are no implementations of `globalLock`, `stickySessionEncoder`, `publicKeyStorage` outside of the Infinispan-implementations. These SPIs are also not part of `Datastore` which makes the handling kind of awkward.

- `globalLock` should be no longer necessary, the SPI was removed completely, maybe we can consider backport to 23.0 if it bothers you.
- `stickySessionEncoder` the current implementation is Infinispan specific and needs to be replaced by something similar to what [we had in Map storage](https://github.com/keycloak/keycloak/blob/release/22.0/server-spi-private/src/main/java/org/keycloak/sessions/DisabledStickySessionEncoderProvider.java). This is something we may want to include by default in Keycloak codebase as not everyone is using Infinispan sticky session provider. 
- `publicKeyStorage` this provider is for caching keys from different sources. Again, we provide only Infinispan implementation and you need to rewrite it if you want to avoid Infinispan. 
 
> * In order to disable Infinispan, one has to disable two `connectionsInfinispan`-implementations: `default` and `quarkus`

Sounds correct to me

> * We basically need to override providers by using the same ID and a higher order to prevent the original implementations (LegacyDatastore, Infinispan-implementations, ...) from registering themselves as event-listeners. This seems like a fundamental problem when overriding provider-implementations since you not only have to make sure to that your provider is used but you also have to prevent that `init()` is called for other provider-implementations you are trying to override...

Keycloak initialize all enabled factories so if you want to avoid initialization of some you need to disable them in config.
 
> The biggest problem seems to be that `LegacyDatastoreProvider` is referenced at multiple points in the code in order to use the `*localStorage`- / `*storageManager`-methods which are not part of `DatastoreProvider` and only exist in `LegacyDatastoreProvider`. Thus we opted to extend `LegacyDatastoreProvider` instead of implementing `DatastoreProvider`. See also #25572 for an attempt to fix this.

This is something we have to revisit. We removed the `userLocalStorage` and `userStorageManager` because map storage do not need them. However, getting back to legacy store means these methods are needed again. I will start a discussion within the team and update this discussion afterwards.

**Reply by opdt** [reply]
Hi @mhajas,

> If I understand correctly, you are also disabling the caching layer, is it correct? The current store is quite coupled with the caching layer and it is not properly tested to run it without it.

Yes and we are fully aware that it is not officially supported. 

> On the other hand, I support the idea of having it more flexible so we would be happy to see how this setup works for you, especially about the performance.

We did this out of necessity but we found after more than a year of using Keycloak that way, that it simplifies operations for us and performance is comparable to Infinispan. We are currently running load tests and if you are interested we could provide you a more comprehensive summary of what we are doing and how it works for us (maybe next Keycloak Dev Day?)

Please note #25830 for a suggestion to simplify replacing datastore and caching layer (you do not have to support this officially, just treat it as one of the many undocumented extension points).

**Reply by mhajas** [reply]
@opdt I started a new discussion for the issue around `UserStoragePrivateUtil`: https://github.com/keycloak/keycloak/discussions/25898
